### PR TITLE
STOR-1819: Add OpenShift specific CSI certification test

### PR DESCRIPTION
**Pull request by jsafrane**
Add a new OpenShift-specific test to `openshift/csi` suite that test a larger number of volumes used in a pod in sequence (or in manageable chunks). There was a 3rd party CSI driver that overflew a kernel counter after 256th attached volume.

This PR introduces OpenShift specific configuration manifest that configures OCP specific tests. There is just one test today, more can come in the future.

Usage:

```
cat <<EOF >ocp-specific-manifest.yaml
Driver: disk.csi.azure.com
LUNStressTest:
  Timeout: "40m"
  PodsTotal: 260
EOF

export TEST_OCP_CSI_DRIVER_FILES=ocp-specific-manifest.yaml
export TEST_CSI_DRIVER_FILES=upstream-test-manifest.yaml
openshift-tests run openshift/csi 
```

The test itself selects a random schedulable node, and creates configured nr. of Pods in parallel (260 by default). The test expects that the CSI driver reports correct attachment limit and the Kubernetes scheduler respects it.
We know [the scheduler can put more volumes to a node than the CSI driver can handle](https://github.com/kubernetes/kubernetes/issues/126502), however, we expect CSI drivers to handle it (and e.g. return errors).

### CODE REVIEW

**Comment by gnufied** [comment]
I was running into: 

```
error: failed to load OCP manifest from "./ocp-specific-manifest.yaml": ./ocp-specific-manifest.yaml: json: cannot unmarshal string into Go struct field LUNStressTestConfig.LUNStressTest.Timeout of type time.Du
ration
```

File: test/extended/storage/csi/scsi_overflow.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		NS:            f.Namespace.Name,
+		PVCs:          []*corev1.PersistentVolumeClaim{pvc},
+		NodeSelection: e2epod.NodeSelection{Name: nodeName},
+		Command:       "ls -la " + e2epod.VolumeMountPath1,
+	}
+	pod, err := e2epod.MakeSecPod(podConfig)
+	// Make the pod name nicer to users (it has random uuid otherwise)
+	pod.Name = fmt.Sprintf("pod-%d", podNumber)
+	e2e.ExpectNoError(err, "preparing pod %d", podNumber)
+	err = e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, pod.Name, pod.Namespace, f.Timeouts.PodDelete)
```
Comment (gnufied) [inline test/extended/storage/csi/scsi_overflow.go:145]: Why this particular line before pod creation? Also, we are ignoring the returned error.

Comment (jsafrane) [reply test/extended/storage/csi/scsi_overflow.go]: This is some leftover, removed.

**Comment by jsafrane** [comment]
I fixed parsing of the timeout, apparently it must be a string and not a Duration.

File: pkg/clioptions/clusterdiscovery/csi.go [hunk 2]
```diff
@@ -6,23 +6,51 @@ import (
...
+	"github.com/openshift/origin/test/extended/storage/csi"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/kubernetes/test/e2e/framework/testfiles"
 	"k8s.io/kubernetes/test/e2e/storage/external"
+	"sigs.k8s.io/yaml"
 )
 
 const (
 	CSIManifestEnvVar = "TEST_CSI_DRIVER_FILES"
+	OCPManifestEnvVar = "TEST_OCP_DRIVER_FILES"
```
[the remark this answers is not in the document]
Comment (bertinatto) [reply pkg/clioptions/clusterdiscovery/csi.go:18]: Whatever the name for the env var we choose, we need to update the doc here: https://github.com/openshift/origin/blob/master/pkg/testsuites/standard_suites.go#L231

File: pkg/clioptions/clusterdiscovery/csi.go [hunk 3]
```diff
@@ -6,23 +6,51 @@ import (
...
 )
 
 // Initialize openshift/csi suite, i.e. define CSI tests from TEST_CSI_DRIVER_FILES.
 func initCSITests(dryRun bool) error {
-	manifestList := os.Getenv(CSIManifestEnvVar)
-	if manifestList != "" {
-		manifests := strings.Split(manifestList, ",")
+	ocpDrivers := sets.New[string]()
```
Comment (bertinatto) [inline pkg/clioptions/clusterdiscovery/csi.go:23]: BTW, I think we're not using `dryRun` anywhere.

File: test/extended/storage/csi/csi.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		LUNStressTest: &LUNStressTestConfig{
+			PodsTotal: 260,
+			Timeout:   DefaultLUNStressTestTimeout,
+		},
+	}
+	if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), bytes, cfg); err != nil {
+		return "", fmt.Errorf("%s: %w", filename, err)
+	}
+
+	testsuites.CSISuites = append(testsuites.CSISuites, initSCSILUNOverflowCSISuite(cfg.LUNStressTest))
```
Comment (bertinatto) [inline test/extended/storage/csi/csi.go:70]: nit: someone unfamiliar with the upstream test suite might be wondering what we're doing here. I'd add  a comment explaining why we're appending to `CSISuites`.

File: test/extended/storage/csi/csi.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	// Select sane defaults
+	cfg := &OpenShiftCSIDriverConfig{
+		LUNStressTest: &LUNStressTestConfig{
+			PodsTotal: 260,
+			Timeout:   DefaultLUNStressTestTimeout,
+		},
+	}
+	if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), bytes, cfg); err != nil {
+		return "", fmt.Errorf("%s: %w", filename, err)
+	}
```
Comment (bertinatto) [inline test/extended/storage/csi/csi.go:68]: Just to be sure: are we OK with not doing any form of validation here (like upstream)?

File: pkg/clioptions/clusterdiscovery/csi.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/clioptions/clusterdiscovery/csi.go]: Changed to TEST_CSI_DRIVER_FILES_FOR_OCP + updated the suite description.

File: pkg/clioptions/clusterdiscovery/csi.go [hunk 6]
```diff
@@ -32,5 +60,34 @@ func initCSITests(dryRun bool) error {
...
+		DriverInfo struct {
+			Name string
+		}
+	}
+	manifest := &upstreamManifest{}
+	err = yaml.Unmarshal(bytes, manifest)
+	if err != nil {
+		return "", err
+	}
+	return manifest.DriverInfo.Name, nil
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/clioptions/clusterdiscovery/csi.go:94]: Added error on empty driver name

File: test/extended/storage/csi/csi.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	// How long to wait for all Pods to start. 40 minutes by default.
+	Timeout time.Duration
+}
+
+// runtime.DecodeInto needs a runtime.Object but doesn't do any
+// deserialization of it and therefore none of the methods below need
+// an implementation.
+var _ runtime.Object = &OpenShiftCSIDriverConfig{}
+
+const DefaultLUNStressTestTimeout = 40 * time.Minute
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply test/extended/storage/csi/csi.go]: Moved up

File: test/extended/storage/csi/csi.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+func AddOpenShiftCSITests(filename string) (string, error) {
+	bytes, err := os.ReadFile(filename)
+	if err != nil {
+		return "", err
+	}
+
+	// Select sane defaults
+	cfg := &OpenShiftCSIDriverConfig{
+		LUNStressTest: &LUNStressTestConfig{
+			PodsTotal: 260,
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply test/extended/storage/csi/csi.go]: Added a constant on the top.

File: test/extended/storage/csi/csi.go (hunk 4 above)
Comment (jsafrane) [reply test/extended/storage/csi/csi.go:80]: I added a comment there... but it's complex

File: test/extended/storage/csi/csi.go (hunk 5 above)
Comment (jsafrane) [reply test/extended/storage/csi/csi.go:74]: yes, no validation. Added a comment about it.

File: pkg/clioptions/clusterdiscovery/csi.go (hunk 3 above)
Comment (jsafrane) [reply pkg/clioptions/clusterdiscovery/csi.go:23]: Removed as a new commit.

**Review by bertinatto — COMMENTED** [review]
/lgtm

**Review by soltysh — APPROVED** [review]
/approve

**Comment by jsafrane** [comment]
/label px-approved
/label docs-approved
/label qe-approved
This will be tested in CI in https://github.com/openshift/release/pull/55204.

**Pull request merged**